datalake/backend/processing/check_files_checksums/task.py

```diff
@@ -1,8 +1,46 @@
 #!/usr/bin/env python3
-
 import sys
+from argparse import ArgumentParser
+from typing import Callable
+
+import boto3
+from botocore.response import StreamingBody
+from multihash import FUNCS, decode
+
+CHUNK_SIZE = 1024
+
+
+def validate_url_multihash(
+    url: str, hex_multihash: str, url_reader: Callable[[str], StreamingBody]
+) -> bool:
+    url_stream = url_reader(url)
+    checksum_function_code = int(hex_multihash[:2], 16)
+    checksum_function = FUNCS[checksum_function_code]
+
+    file_digest = checksum_function()
+    for chunk in url_stream.iter_chunks(chunk_size=CHUNK_SIZE):
+        file_digest.update(chunk)
+
+    return file_digest.digest() == decode(bytes.fromhex(hex_multihash))
+
+
+def parse_arguments():
```

Comment:
This script is not going to receive any of `file-url` or `hex-multihash` arguments. The only information it will have is metadata URL (`--metadata-url`) and index of file to check as env variables (`BATCH_JOB_FIRST_ITEM_INDEX` and `AWS_BATCH_JOB_ARRAY_INDEX`). See: Check Data Items Checksums (Multiple Parallel AWS Batch Jobs)
 in https://confluence.linz.govt.nz/display/LGDL/Dataset+Version+Endpoint

Reply:
That comes back to [this issue](https://github.com/linz/geospatial-data-lake/issues/164#issuecomment-754917267): We really should not have to parse the entire set of metadata more than once. Possibly this means that the metadata check should save a list of all the unique links + checksums it encounters to the DB.

Reply:
As discussed we should probably resolve the storage issue before tweaking this job. To summarise the discussion:

- The `check_stac_metadata` job seems like the right place to pull out the list of links + checksums to avoid reprocessing the metadata files.
- The list should be stored in a way that makes it efficient to get entries N through N+M for batch processing.
- We're not sure whether S3, DynamoDB or something else is the optimal solution in terms of complexity/flexibility. For example, does DynamoDB have a low limit on the number of rows in a table?

Reply:
Nice one!